### PR TITLE
Limit the resutls objects SearchContext creates

### PR DESCRIPTION
**Pull request by not-napoleon**
Relates to https://github.com/elastic/elasticsearch/pull/94023

While working on releasing the aggregation context when we clean up `QuerySearchResult`, I ran into a problem that we were creating query search results in places we didn't need them.  This is at least in part because `DefaultSearchContext` always creates results objects for DFS, Query, and Fetch phases, even if we will only need one of those.  This PR alters that behavior so callers must now request the results type they intend to use.

I'm adding this as a separate PR because it's a fairly deep change that I'd like to debug and review independent of the changes to `QuerySearchResult`, but ultimately this is still in service of the above linked refactor.

### CODE REVIEW

**Review by benwtrent — COMMENTED** [review]
🤔 this design almost makes me think we should make search context `abstract` and enforce a `DFSSearchContext`, etc...

I will review deeply soon.

**Comment by not-napoleon** [comment]
@benwtrent `SearchContext` is abstract.  The two (main) implementations are `DefaultSearchContext` and `FilterSearchContext`.  Also, it turns out there's a lot of places we need both Query and Fetch results.  I'm working through those now.

**Review by martijnvg — COMMENTED** [review]
This change looks good to me. 
I just don't yet understand why `30_inner_hits/Test two sub-queries with only one having inner_hits` test fails.

**Comment by martijnvg** [comment]
> I just don't yet understand why 30_inner_hits/Test two sub-queries with only one having inner_hits test fails.

I think the `getTotalHits()` and `getMaxScore()` methods need to be overwritten in `SubSearchContext`.

**Review by martijnvg — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/search/fetch/QueryFetchSearchResult.java [hunk 1]
```diff
@@ -24,6 +24,7 @@ public final class QueryFetchSearchResult extends SearchPhaseResult {
 
     public QueryFetchSearchResult(StreamInput in) throws IOException {
         super(in);
+        // TODO: Delegate refcounting to QuerySearchResult
```
Comment (jdconrad) [inline server/src/main/java/org/elasticsearch/search/fetch/QueryFetchSearchResult.java:27]: Is it worth adding the other PR number this references?

**Review by jdconrad — APPROVED** [review]
LGTM! Thanks again for pointing this out. It shouldn't have much overlap with (https://github.com/elastic/elasticsearch/pull/93396) outside of a few additional methods for an overridden `SearchContext`.

File: server/src/main/java/org/elasticsearch/search/fetch/QueryFetchSearchResult.java (hunk 1 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/fetch/QueryFetchSearchResult.java:27]: yeah, that's a good idea.  I'll add it in.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Pull request merged**